src/System.IO.Compression.Brotli/System/IO/Compression/Brotli.cs

```diff
@@ -86,17 +183,42 @@ public static TransformationStatus Decompress(ReadOnlySpan<byte> source, Span<by
                 {
                     bufIn = new IntPtr(inBytes);
                     bufOut = new IntPtr(outBytes);
-                    nuint written = (nuint)destination.Length;
-                    nuint consumed = (nuint)source.Length;
-                    BrotliDecoderResult res = BrotliNative.BrotliDecoderDecompress(ref consumed, bufIn, ref written, bufOut);
-                    if (res == BrotliDecoderResult.Success)
-                    {
-                        bytesWritten = (int)written;
-                        bytesConsumed = (int)consumed;
-                    }
-                    return GetTransformationStatusFromBrotliDecoderResult(res);
+                    nuint written = (nuint)bytesWritten;
+                    nuint consumed = (nuint)bytesConsumed;
+                    LastDecoderResult = BrotliNative.BrotliDecoderDecompressStream(state.BrotliNativeState, ref consumed, ref bufIn, ref written, ref bufOut, out nuint totalOut);
+                    bytesWritten = (int)((nuint)destination.Length - written);
+                    bytesConsumed = (int)consumed;
+                }
+                if (LastDecoderResult == BrotliDecoderResult.NeedsMoreInput)
+                {
+                    return TransformationStatus.NeedMoreSourceData;
+                }
+                else if (LastDecoderResult == BrotliDecoderResult.NeedsMoreOutput)
+                {
+                    return TransformationStatus.DestinationTooSmall;
+                }
+                else
+                {
+                    endOfStream = true;
                 }
+
+                if (endOfStream && !BrotliNative.BrotliDecoderIsFinished(state.BrotliNativeState))
+                {
+                    errorDetected = true;
```

Comment:
Similarly, I don't think we need this. You can just use `!BrotliNative.BrotliDecoderIsFinished(state.BrotliNativeState))` at line 209. Let me know if I am mistaken.